### PR TITLE
Add support for TabNet

### PR DESCRIPTION
**Pull request by pfistfl**
Requested in #7 

This is a very generic POC, I think this will require a little more thinking, to make stuff beautiful.

- We now have transforms that transform from `Task` to `tf.data | matrix`.
  This can be overwritten and then allows for more flexibility.
- There is now a unit test for TabNet, that shows how TabNet could be used with `classif.keras`.

### CODE REVIEW

File: tests/testthat/test_tabnet.R [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  tsk = mlr_tasks$get("iris")
+  make_feature_column = function(id, type) {
+    if (type == "numeric") tf$feature_column$numeric_column(id)
+    else tf$feature_column$categorical_column_with_vocabulary_list(id, tsk$levels(id)[[1]])
+  }
+  feature_columns = pmap(.f = make_feature_column, .x = tsk$feature_types)
+ 
+  # Define and compile model
+  tabnet = import("tabnet") 
+  model = tabnet$TabNetClassification(feature_columns, num_classes=3,
```
Comment (JackyP) [inline tests/testthat/test_tabnet.R]: This is awesome! 

Just a note to say that thanks to a quick fix (sorry - I've been unwell this week!) to https://github.com/titu1994/tf-TabNet/issues/4 there's now a version "0.1.4.1" In this version the class has been renamed to "TabNetClassifier".

File: tests/testthat/test_tabnet.R [hunk 2]
```diff
@@ -0,0 +1,50 @@
+context("tabnet")
+
+test_that("test tabnet", {
+  library("reticulate")
+  library("tensorflow")
+  # keras::install_keras(extra_packages = c("tensorflow-hub", "tabnet==0.1.3"))
```
Comment (JackyP) [inline tests/testthat/test_tabnet.R]: There's now a 0.1.4.1 per below

File: tests/testthat/test_tabnet.R [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  lrn$param_set$values$validation_split = NULL # Does not work with tf.data
+
+  # We overwrite task -> dataset transformers here to stay flexible.
+  lrn$set_transform("x", function(task, pars) {
+    x = lapply(task$feature_names, function(x) { as.matrix(task$data(cols = x))})
+    names(x) = task$feature_names
+    return(x)
+  })
+  lrn$train(tsk)
+  expect_class(lrn$model$model, "tabnet.tabnet.TabNetClassification")
```
Comment (JackyP) [inline tests/testthat/test_tabnet.R]: This would also change to tabnet.tabnet.TabNetClassifier

**Review by JackyP — COMMENTED** [review]
This change generalises the x and y transform which allows dictionary-style x for multi-input models. 

The y transform may potentially be handy for multi-output models. Currently it does not seem like mlr3 is designed for multi-task regression, but the logic makes sense. A test case for the y_transform may be helpful for coverage.

**Comment by JackyP** [comment]
Thinking about this a bit more - is the flexibility of the x_transform and y_transform needed, or should it be more constrained?

The reference docs only have a limited number of input formats into fit

https://keras.rstudio.com/reference/fit.html 
> Vector, matrix, or array of training data (or list if the model has multiple inputs). If all inputs in the model are named, you can also pass a list mapping input names to data. 

The default case handles the matrix-like, and the tabnet case is the list-like

Could the function just check 
```
length(model$input_names)==1
```
and use that to pick the matrix-like or list-like formats by default I wonder?

**Comment by pfistfl** [comment]
Hey, thanks for the comments.

I was hoping I can generalize that! 
I will try to come up with a prototype at the end of the week.

> The y transform may potentially be handy for multi-output models. Currently, it does not seem like mlr3 is designed for multi-task regression, but the logic makes sense. A test case for the y_transform may be helpful for coverage.
Yes, and I think this will come soon(ish). Will do!

**Comment by pfistfl** [comment]
Hey,
I did a major refactoring. This lets me do a lot of things in way less code.
(I hope stuff does not get to complicated as a result).

This now includes tabnet and regression.

**Comment by pfistfl** [comment]
@JackyP 
Any thoughts? I feel I might have made things more complicated than need be. 
If you feel this is alright, we can merge this and try to get PR #12  merged aswell?

**Comment by JackyP** [comment]
It does feel like it could be simplified a little, but it was not too hard to work with. Keen to hear your views on #12 as well.

**Pull request merged**